# Лабораторная работа №1
# Функция Капрекара - K(i,k,r)

### Разделение числа на состовляющие цифры:

In [1]:
def list_from_num(number: int, k: int, r: int):
    list_of_digits = list()
    num=number
    for m in range(k):
        num, tmp = divmod(num, r)
        list_of_digits.append(tmp)
    return list_of_digits

#Тестовые примеры на разные системы счисления
print (list_from_num(1234, 6, 10))
print(list_from_num(10, 4, 2))
print(list_from_num(10,8,2))
print (list_from_num(10, 2, 8))
print(list_from_num(16, 2, 16))
    

[4, 3, 2, 1, 0, 0]
[0, 1, 0, 1]
[0, 1, 0, 1, 0, 0, 0, 0]
[2, 1]
[0, 1]


### Сборка числа из элементов(цифр) списка:

In [2]:
def num_from_list(l:list, k:int, r:int):
    num=0
    for i in range(len(l)):
        num=num*r+l[i];
    return num

#Тестовые примеры на разные системы счисления
print(num_from_list([4, 3, 2, 1, 0, 0], 6,10))
print(num_from_list([1, 1, 0, 1], 4,2))
print(num_from_list([2,1],2,8))
print(num_from_list([1,0],2,16))   
    

432100
13
17
16


### Функция Капрекар:
number - число  
k - разрядность числа  
r - система счисления  

In [3]:
def kaprekar(number:int, k:int, r:int):
    list_of_digits = list_from_num(number,k,r)
    list_of_digits.sort()
    min = num_from_list(list_of_digits, k, r)
    list_of_digits.sort(reverse=True)
    max = num_from_list(list_of_digits, k, r)
    new_list = list(str(max - min))
    new = int("".join(map(str, new_list)))
    return new

print(kaprekar(1000, 4, 10))
print(kaprekar(4196, 4, 10))
print(kaprekar(0, 4, 10))

999
8172
0


In [4]:
def magic_number(num:int, k:int, r:int):
    new_num=kaprekar(num, k, r)
    if new_num == num:
        return num
    else:
        return -1

#Считается ли число магическим?
print(magic_number(6174, 4,10))
print(magic_number(3030,4, 10))
print(magic_number(549945, 6, 10))
    

6174
-1
549945


## Функция, возвращающая список всех магический числел в заданной сиситеме счисления(r), с заданной разрядностью(k):

In [93]:
def magic_numbers(k: int, r: int):
    new = 0
    magic_nums = list()
    exist_nums = list()
    s='9'*k
    for i in range(int(s)+1):
        #print(i, end='')
        while True:
            new = kaprekar(i, k, r)
            #print("---> %d" % new, end='')
            if new == i:
                if new in magic_nums:
                    break
                magic_nums.append(new)
                #print("\n")
                break
            if new in exist_nums:
                #print("--->exist")
                break
            i = new
            exist_nums.append(new)
    return magic_nums
    #print("\nMagic numbers:",magic_nums)  


In [89]:
#Проверка корректной работы функции
print("Magic Numbers:",magic_numbers(4,10))

0---> 0

1---> 999---> 8991---> 8082---> 8532---> 6174---> 6174

2---> 1998---> 8082--->exist
3---> 2997---> 7173---> 6354---> 3087---> 8352---> 6174--->exist
4---> 3996---> 6264---> 4176---> 6174--->exist
5---> 4995---> 5355---> 1998--->exist
6---> 5994---> 5355--->exist
7---> 6993---> 6264--->exist
8---> 7992---> 7173--->exist
9---> 8991--->exist
10---> 999--->exist
11---> 1089---> 9621---> 8352--->exist
12---> 2088---> 8532--->exist
13---> 3087--->exist
14---> 4086---> 8172---> 7443---> 3996--->exist
15---> 5085---> 7992--->exist
16---> 6084---> 8172--->exist
17---> 7083---> 8352--->exist
18---> 8082--->exist
19---> 9081---> 9621--->exist
20---> 1998--->exist
21---> 2088--->exist
22---> 2178---> 7443--->exist
23---> 3177---> 6354--->exist
24---> 4176--->exist
25---> 5175---> 5994--->exist
26---> 6174--->exist
27---> 7173--->exist
28---> 8172--->exist
29---> 9171---> 8532--->exist
30---> 2997--->exist
31---> 3087--->exist
32---> 3177--->exist
33---> 3267---> 5265---> 3996--->exist
34

484---> 7992--->exist
485---> 8082--->exist
486---> 8172--->exist
487---> 8262--->exist
488---> 8352--->exist
489---> 9351--->exist
490---> 9351--->exist
491---> 9261--->exist
492---> 9171--->exist
493---> 9081--->exist
494---> 8991--->exist
495---> 9081--->exist
496---> 9171--->exist
497---> 9261--->exist
498---> 9351--->exist
499---> 9441--->exist
500---> 4995--->exist
501---> 5085--->exist
502---> 5175--->exist
503---> 5265--->exist
504---> 5355--->exist
505---> 5445--->exist
506---> 6444--->exist
507---> 7443--->exist
508---> 8442--->exist
509---> 9441--->exist
510---> 5085--->exist
511---> 4995--->exist
512---> 5085--->exist
513---> 5175--->exist
514---> 5265--->exist
515---> 5355--->exist
516---> 6354--->exist
517---> 7353--->exist
518---> 8352--->exist
519---> 9351--->exist
520---> 5175--->exist
521---> 5085--->exist
522---> 4995--->exist
523---> 5085--->exist
524---> 5175--->exist
525---> 5265--->exist
526---> 6264--->exist
527---> 7263--->exist
528---> 8262--->exist
529---> 92

987---> 9081--->exist
988---> 8991--->exist
989---> 9081--->exist
990---> 9801--->exist
991---> 9711--->exist
992---> 9621--->exist
993---> 9531--->exist
994---> 9441--->exist
995---> 9351--->exist
996---> 9261--->exist
997---> 9171--->exist
998---> 9081--->exist
999---> 8991--->exist
1000---> 999--->exist
1001---> 1089--->exist
1002---> 2088--->exist
1003---> 3087--->exist
1004---> 4086--->exist
1005---> 5085--->exist
1006---> 6084--->exist
1007---> 7083--->exist
1008---> 8082--->exist
1009---> 9081--->exist
1010---> 1089--->exist
1011---> 999--->exist
1012---> 1998--->exist
1013---> 2997--->exist
1014---> 3996--->exist
1015---> 4995--->exist
1016---> 5994--->exist
1017---> 6993--->exist
1018---> 7992--->exist
1019---> 8991--->exist
1020---> 2088--->exist
1021---> 1998--->exist
1022---> 2088--->exist
1023---> 3087--->exist
1024---> 4086--->exist
1025---> 5085--->exist
1026---> 6084--->exist
1027---> 7083--->exist
1028---> 8082--->exist
1029---> 9081--->exist
1030---> 3087--->exist
103

1450---> 5265--->exist
1451---> 4266--->exist
1452---> 4176--->exist
1453---> 4086--->exist
1454---> 3996--->exist
1455---> 4086--->exist
1456---> 5085--->exist
1457---> 6084--->exist
1458---> 7083--->exist
1459---> 8082--->exist
1460---> 6264--->exist
1461---> 5265--->exist
1462---> 5175--->exist
1463---> 5085--->exist
1464---> 4995--->exist
1465---> 5085--->exist
1466---> 5175--->exist
1467---> 6174--->exist
1468---> 7173--->exist
1469---> 8172--->exist
1470---> 7263--->exist
1471---> 6264--->exist
1472---> 6174--->exist
1473---> 6084--->exist
1474---> 5994--->exist
1475---> 6084--->exist
1476---> 6174--->exist
1477---> 6264--->exist
1478---> 7263--->exist
1479---> 8262--->exist
1480---> 8262--->exist
1481---> 7263--->exist
1482---> 7173--->exist
1483---> 7083--->exist
1484---> 6993--->exist
1485---> 7083--->exist
1486---> 7173--->exist
1487---> 7263--->exist
1488---> 7353--->exist
1489---> 8352--->exist
1490---> 9261--->exist
1491---> 8262--->exist
1492---> 8172--->exist
1493---> 80

1930---> 9171--->exist
1931---> 8172--->exist
1932---> 8082--->exist
1933---> 7992--->exist
1934---> 8082--->exist
1935---> 8172--->exist
1936---> 8262--->exist
1937---> 8352--->exist
1938---> 8442--->exist
1939---> 8532--->exist
1940---> 9261--->exist
1941---> 8262--->exist
1942---> 8172--->exist
1943---> 8082--->exist
1944---> 7992--->exist
1945---> 8082--->exist
1946---> 8172--->exist
1947---> 8262--->exist
1948---> 8352--->exist
1949---> 8442--->exist
1950---> 9351--->exist
1951---> 8352--->exist
1952---> 8262--->exist
1953---> 8172--->exist
1954---> 8082--->exist
1955---> 7992--->exist
1956---> 8082--->exist
1957---> 8172--->exist
1958---> 8262--->exist
1959---> 8352--->exist
1960---> 9441--->exist
1961---> 8442--->exist
1962---> 8352--->exist
1963---> 8262--->exist
1964---> 8172--->exist
1965---> 8082--->exist
1966---> 7992--->exist
1967---> 8082--->exist
1968---> 8172--->exist
1969---> 8262--->exist
1970---> 9531--->exist
1971---> 8532--->exist
1972---> 8442--->exist
1973---> 83

2402---> 3996--->exist
2403---> 4086--->exist
2404---> 4176--->exist
2405---> 5175--->exist
2406---> 6174--->exist
2407---> 7173--->exist
2408---> 8172--->exist
2409---> 9171--->exist
2410---> 4086--->exist
2411---> 3087--->exist
2412---> 2997--->exist
2413---> 3087--->exist
2414---> 3177--->exist
2415---> 4176--->exist
2416---> 5175--->exist
2417---> 6174--->exist
2418---> 7173--->exist
2419---> 8172--->exist
2420---> 3996--->exist
2421---> 2997--->exist
2422---> 1998--->exist
2423---> 2088--->exist
2424---> 2178--->exist
2425---> 3177--->exist
2426---> 4176--->exist
2427---> 5175--->exist
2428---> 6174--->exist
2429---> 7173--->exist
2430---> 4086--->exist
2431---> 3087--->exist
2432---> 2088--->exist
2433---> 1998--->exist
2434---> 2088--->exist
2435---> 3087--->exist
2436---> 4086--->exist
2437---> 5085--->exist
2438---> 6084--->exist
2439---> 7083--->exist
2440---> 4176--->exist
2441---> 3177--->exist
2442---> 2178--->exist
2443---> 2088--->exist
2444---> 1998--->exist
2445---> 29

2916---> 8352--->exist
2917---> 8442--->exist
2918---> 8532--->exist
2919---> 8622--->exist
2920---> 8991--->exist
2921---> 7992--->exist
2922---> 6993--->exist
2923---> 7083--->exist
2924---> 7173--->exist
2925---> 7263--->exist
2926---> 7353--->exist
2927---> 7443--->exist
2928---> 7533--->exist
2929---> 7623--->exist
2930---> 9081--->exist
2931---> 8082--->exist
2932---> 7083--->exist
2933---> 6993--->exist
2934---> 7083--->exist
2935---> 7173--->exist
2936---> 7263--->exist
2937---> 7353--->exist
2938---> 7443--->exist
2939---> 7533--->exist
2940---> 9171--->exist
2941---> 8172--->exist
2942---> 7173--->exist
2943---> 7083--->exist
2944---> 6993--->exist
2945---> 7083--->exist
2946---> 7173--->exist
2947---> 7263--->exist
2948---> 7353--->exist
2949---> 7443--->exist
2950---> 9261--->exist
2951---> 8262--->exist
2952---> 7263--->exist
2953---> 7173--->exist
2954---> 7083--->exist
2955---> 6993--->exist
2956---> 7083--->exist
2957---> 7173--->exist
2958---> 7263--->exist
2959---> 73

3395---> 6174--->exist
3396---> 6264--->exist
3397---> 6354--->exist
3398---> 6444--->exist
3399---> 6534--->exist
3400---> 4266--->exist
3401---> 4176--->exist
3402---> 4086--->exist
3403---> 3996--->exist
3404---> 4086--->exist
3405---> 5085--->exist
3406---> 6084--->exist
3407---> 7083--->exist
3408---> 8082--->exist
3409---> 9081--->exist
3410---> 4176--->exist
3411---> 3177--->exist
3412---> 3087--->exist
3413---> 2997--->exist
3414---> 3087--->exist
3415---> 4086--->exist
3416---> 5085--->exist
3417---> 6084--->exist
3418---> 7083--->exist
3419---> 8082--->exist
3420---> 4086--->exist
3421---> 3087--->exist
3422---> 2088--->exist
3423---> 1998--->exist
3424---> 2088--->exist
3425---> 3087--->exist
3426---> 4086--->exist
3427---> 5085--->exist
3428---> 6084--->exist
3429---> 7083--->exist
3430---> 3996--->exist
3431---> 2997--->exist
3432---> 1998--->exist
3433---> 999--->exist
3434---> 1089--->exist
3435---> 2088--->exist
3436---> 3087--->exist
3437---> 4086--->exist
3438---> 508

3903---> 8991--->exist
3904---> 9081--->exist
3905---> 9171--->exist
3906---> 9261--->exist
3907---> 9351--->exist
3908---> 9441--->exist
3909---> 9531--->exist
3910---> 9171--->exist
3911---> 8172--->exist
3912---> 8082--->exist
3913---> 7992--->exist
3914---> 8082--->exist
3915---> 8172--->exist
3916---> 8262--->exist
3917---> 8352--->exist
3918---> 8442--->exist
3919---> 8532--->exist
3920---> 9081--->exist
3921---> 8082--->exist
3922---> 7083--->exist
3923---> 6993--->exist
3924---> 7083--->exist
3925---> 7173--->exist
3926---> 7263--->exist
3927---> 7353--->exist
3928---> 7443--->exist
3929---> 7533--->exist
3930---> 8991--->exist
3931---> 7992--->exist
3932---> 6993--->exist
3933---> 5994--->exist
3934---> 6084--->exist
3935---> 6174--->exist
3936---> 6264--->exist
3937---> 6354--->exist
3938---> 6444--->exist
3939---> 6534--->exist
3940---> 9081--->exist
3941---> 8082--->exist
3942---> 7083--->exist
3943---> 6084--->exist
3944---> 5994--->exist
3945---> 6084--->exist
3946---> 61

4510---> 5265--->exist
4511---> 4266--->exist
4512---> 4176--->exist
4513---> 4086--->exist
4514---> 3996--->exist
4515---> 4086--->exist
4516---> 5085--->exist
4517---> 6084--->exist
4518---> 7083--->exist
4519---> 8082--->exist
4520---> 5175--->exist
4521---> 4176--->exist
4522---> 3177--->exist
4523---> 3087--->exist
4524---> 2997--->exist
4525---> 3087--->exist
4526---> 4086--->exist
4527---> 5085--->exist
4528---> 6084--->exist
4529---> 7083--->exist
4530---> 5085--->exist
4531---> 4086--->exist
4532---> 3087--->exist
4533---> 2088--->exist
4534---> 1998--->exist
4535---> 2088--->exist
4536---> 3087--->exist
4537---> 4086--->exist
4538---> 5085--->exist
4539---> 6084--->exist
4540---> 4995--->exist
4541---> 3996--->exist
4542---> 2997--->exist
4543---> 1998--->exist
4544---> 999--->exist
4545---> 1089--->exist
4546---> 2088--->exist
4547---> 3087--->exist
4548---> 4086--->exist
4549---> 5085--->exist
4550---> 5085--->exist
4551---> 4086--->exist
4552---> 3087--->exist
4553---> 208

4922---> 7173--->exist
4923---> 7083--->exist
4924---> 6993--->exist
4925---> 7083--->exist
4926---> 7173--->exist
4927---> 7263--->exist
4928---> 7353--->exist
4929---> 7443--->exist
4930---> 9081--->exist
4931---> 8082--->exist
4932---> 7083--->exist
4933---> 6084--->exist
4934---> 5994--->exist
4935---> 6084--->exist
4936---> 6174--->exist
4937---> 6264--->exist
4938---> 6354--->exist
4939---> 6444--->exist
4940---> 8991--->exist
4941---> 7992--->exist
4942---> 6993--->exist
4943---> 5994--->exist
4944---> 4995--->exist
4945---> 5085--->exist
4946---> 5175--->exist
4947---> 5265--->exist
4948---> 5355--->exist
4949---> 5445--->exist
4950---> 9081--->exist
4951---> 8082--->exist
4952---> 7083--->exist
4953---> 6084--->exist
4954---> 5085--->exist
4955---> 4995--->exist
4956---> 5085--->exist
4957---> 5175--->exist
4958---> 5265--->exist
4959---> 5355--->exist
4960---> 9171--->exist
4961---> 8172--->exist
4962---> 7173--->exist
4963---> 6174--->exist
4964---> 5175--->exist
4965---> 50

5414---> 3996--->exist
5415---> 4086--->exist
5416---> 5085--->exist
5417---> 6084--->exist
5418---> 7083--->exist
5419---> 8082--->exist
5420---> 5175--->exist
5421---> 4176--->exist
5422---> 3177--->exist
5423---> 3087--->exist
5424---> 2997--->exist
5425---> 3087--->exist
5426---> 4086--->exist
5427---> 5085--->exist
5428---> 6084--->exist
5429---> 7083--->exist
5430---> 5085--->exist
5431---> 4086--->exist
5432---> 3087--->exist
5433---> 2088--->exist
5434---> 1998--->exist
5435---> 2088--->exist
5436---> 3087--->exist
5437---> 4086--->exist
5438---> 5085--->exist
5439---> 6084--->exist
5440---> 4995--->exist
5441---> 3996--->exist
5442---> 2997--->exist
5443---> 1998--->exist
5444---> 999--->exist
5445---> 1089--->exist
5446---> 2088--->exist
5447---> 3087--->exist
5448---> 4086--->exist
5449---> 5085--->exist
5450---> 5085--->exist
5451---> 4086--->exist
5452---> 3087--->exist
5453---> 2088--->exist
5454---> 1089--->exist
5455---> 999--->exist
5456---> 1998--->exist
5457---> 2997

5946---> 5085--->exist
5947---> 5175--->exist
5948---> 5265--->exist
5949---> 5355--->exist
5950---> 8991--->exist
5951---> 7992--->exist
5952---> 6993--->exist
5953---> 5994--->exist
5954---> 4995--->exist
5955---> 3996--->exist
5956---> 4086--->exist
5957---> 4176--->exist
5958---> 4266--->exist
5959---> 4356--->exist
5960---> 9081--->exist
5961---> 8082--->exist
5962---> 7083--->exist
5963---> 6084--->exist
5964---> 5085--->exist
5965---> 4086--->exist
5966---> 3996--->exist
5967---> 4086--->exist
5968---> 4176--->exist
5969---> 4266--->exist
5970---> 9171--->exist
5971---> 8172--->exist
5972---> 7173--->exist
5973---> 6174--->exist
5974---> 5175--->exist
5975---> 4176--->exist
5976---> 4086--->exist
5977---> 3996--->exist
5978---> 4086--->exist
5979---> 4176--->exist
5980---> 9261--->exist
5981---> 8262--->exist
5982---> 7263--->exist
5983---> 6264--->exist
5984---> 5265--->exist
5985---> 4266--->exist
5986---> 4176--->exist
5987---> 4086--->exist
5988---> 3996--->exist
5989---> 40

6443---> 2997--->exist
6444---> 1998--->exist
6445---> 2088--->exist
6446---> 2178--->exist
6447---> 3177--->exist
6448---> 4176--->exist
6449---> 5175--->exist
6450---> 6084--->exist
6451---> 5085--->exist
6452---> 4086--->exist
6453---> 3087--->exist
6454---> 2088--->exist
6455---> 1998--->exist
6456---> 2088--->exist
6457---> 3087--->exist
6458---> 4086--->exist
6459---> 5085--->exist
6460---> 6174--->exist
6461---> 5175--->exist
6462---> 4176--->exist
6463---> 3177--->exist
6464---> 2178--->exist
6465---> 2088--->exist
6466---> 1998--->exist
6467---> 2997--->exist
6468---> 3996--->exist
6469---> 4995--->exist
6470---> 7173--->exist
6471---> 6174--->exist
6472---> 5175--->exist
6473---> 4176--->exist
6474---> 3177--->exist
6475---> 3087--->exist
6476---> 2997--->exist
6477---> 3087--->exist
6478---> 4086--->exist
6479---> 5085--->exist
6480---> 8172--->exist
6481---> 7173--->exist
6482---> 6174--->exist
6483---> 5175--->exist
6484---> 4176--->exist
6485---> 4086--->exist
6486---> 39

6967---> 3087--->exist
6968---> 3177--->exist
6969---> 3267--->exist
6970---> 9081--->exist
6971---> 8082--->exist
6972---> 7083--->exist
6973---> 6084--->exist
6974---> 5085--->exist
6975---> 4086--->exist
6976---> 3087--->exist
6977---> 2997--->exist
6978---> 3087--->exist
6979---> 3177--->exist
6980---> 9171--->exist
6981---> 8172--->exist
6982---> 7173--->exist
6983---> 6174--->exist
6984---> 5175--->exist
6985---> 4176--->exist
6986---> 3177--->exist
6987---> 3087--->exist
6988---> 2997--->exist
6989---> 3087--->exist
6990---> 9261--->exist
6991---> 8262--->exist
6992---> 7263--->exist
6993---> 6264--->exist
6994---> 5265--->exist
6995---> 4266--->exist
6996---> 3267--->exist
6997---> 3177--->exist
6998---> 3087--->exist
6999---> 2997--->exist
7000---> 6993--->exist
7001---> 7083--->exist
7002---> 7173--->exist
7003---> 7263--->exist
7004---> 7353--->exist
7005---> 7443--->exist
7006---> 7533--->exist
7007---> 7623--->exist
7008---> 8622--->exist
7009---> 9621--->exist
7010---> 70

7439---> 6264--->exist
7440---> 6993--->exist
7441---> 5994--->exist
7442---> 4995--->exist
7443---> 3996--->exist
7444---> 2997--->exist
7445---> 3087--->exist
7446---> 3177--->exist
7447---> 3267--->exist
7448---> 4266--->exist
7449---> 5265--->exist
7450---> 7083--->exist
7451---> 6084--->exist
7452---> 5085--->exist
7453---> 4086--->exist
7454---> 3087--->exist
7455---> 2997--->exist
7456---> 3087--->exist
7457---> 3177--->exist
7458---> 4176--->exist
7459---> 5175--->exist
7460---> 7173--->exist
7461---> 6174--->exist
7462---> 5175--->exist
7463---> 4176--->exist
7464---> 3177--->exist
7465---> 3087--->exist
7466---> 2997--->exist
7467---> 3087--->exist
7468---> 4086--->exist
7469---> 5085--->exist
7470---> 7263--->exist
7471---> 6264--->exist
7472---> 5265--->exist
7473---> 4266--->exist
7474---> 3267--->exist
7475---> 3177--->exist
7476---> 3087--->exist
7477---> 2997--->exist
7478---> 3996--->exist
7479---> 4995--->exist
7480---> 8262--->exist
7481---> 7263--->exist
7482---> 62

7967---> 2997--->exist
7968---> 3087--->exist
7969---> 3177--->exist
7970---> 8991--->exist
7971---> 7992--->exist
7972---> 6993--->exist
7973---> 5994--->exist
7974---> 4995--->exist
7975---> 3996--->exist
7976---> 2997--->exist
7977---> 1998--->exist
7978---> 2088--->exist
7979---> 2178--->exist
7980---> 9081--->exist
7981---> 8082--->exist
7982---> 7083--->exist
7983---> 6084--->exist
7984---> 5085--->exist
7985---> 4086--->exist
7986---> 3087--->exist
7987---> 2088--->exist
7988---> 1998--->exist
7989---> 2088--->exist
7990---> 9171--->exist
7991---> 8172--->exist
7992---> 7173--->exist
7993---> 6174--->exist
7994---> 5175--->exist
7995---> 4176--->exist
7996---> 3177--->exist
7997---> 2178--->exist
7998---> 2088--->exist
7999---> 1998--->exist
8000---> 7992--->exist
8001---> 8082--->exist
8002---> 8172--->exist
8003---> 8262--->exist
8004---> 8352--->exist
8005---> 8442--->exist
8006---> 8532--->exist
8007---> 8622--->exist
8008---> 8712--->exist
8009---> 9711--->exist
8010---> 80

8450---> 8082--->exist
8451---> 7083--->exist
8452---> 6084--->exist
8453---> 5085--->exist
8454---> 4086--->exist
8455---> 3996--->exist
8456---> 4086--->exist
8457---> 4176--->exist
8458---> 4266--->exist
8459---> 5265--->exist
8460---> 8172--->exist
8461---> 7173--->exist
8462---> 6174--->exist
8463---> 5175--->exist
8464---> 4176--->exist
8465---> 4086--->exist
8466---> 3996--->exist
8467---> 4086--->exist
8468---> 4176--->exist
8469---> 5175--->exist
8470---> 8262--->exist
8471---> 7263--->exist
8472---> 6264--->exist
8473---> 5265--->exist
8474---> 4266--->exist
8475---> 4176--->exist
8476---> 4086--->exist
8477---> 3996--->exist
8478---> 4086--->exist
8479---> 5085--->exist
8480---> 8352--->exist
8481---> 7353--->exist
8482---> 6354--->exist
8483---> 5355--->exist
8484---> 4356--->exist
8485---> 4266--->exist
8486---> 4176--->exist
8487---> 4086--->exist
8488---> 3996--->exist
8489---> 4995--->exist
8490---> 9351--->exist
8491---> 8352--->exist
8492---> 7353--->exist
8493---> 63

8936---> 6174--->exist
8937---> 6084--->exist
8938---> 5994--->exist
8939---> 6084--->exist
8940---> 9351--->exist
8941---> 8352--->exist
8942---> 7353--->exist
8943---> 6354--->exist
8944---> 5355--->exist
8945---> 5265--->exist
8946---> 5175--->exist
8947---> 5085--->exist
8948---> 4995--->exist
8949---> 5085--->exist
8950---> 9261--->exist
8951---> 8262--->exist
8952---> 7263--->exist
8953---> 6264--->exist
8954---> 5265--->exist
8955---> 4266--->exist
8956---> 4176--->exist
8957---> 4086--->exist
8958---> 3996--->exist
8959---> 4086--->exist
8960---> 9171--->exist
8961---> 8172--->exist
8962---> 7173--->exist
8963---> 6174--->exist
8964---> 5175--->exist
8965---> 4176--->exist
8966---> 3177--->exist
8967---> 3087--->exist
8968---> 2997--->exist
8969---> 3087--->exist
8970---> 9081--->exist
8971---> 8082--->exist
8972---> 7083--->exist
8973---> 6084--->exist
8974---> 5085--->exist
8975---> 4086--->exist
8976---> 3087--->exist
8977---> 2088--->exist
8978---> 1998--->exist
8979---> 20

9420---> 9171--->exist
9421---> 8172--->exist
9422---> 7173--->exist
9423---> 7083--->exist
9424---> 6993--->exist
9425---> 7083--->exist
9426---> 7173--->exist
9427---> 7263--->exist
9428---> 7353--->exist
9429---> 7443--->exist
9430---> 9081--->exist
9431---> 8082--->exist
9432---> 7083--->exist
9433---> 6084--->exist
9434---> 5994--->exist
9435---> 6084--->exist
9436---> 6174--->exist
9437---> 6264--->exist
9438---> 6354--->exist
9439---> 6444--->exist
9440---> 8991--->exist
9441---> 7992--->exist
9442---> 6993--->exist
9443---> 5994--->exist
9444---> 4995--->exist
9445---> 5085--->exist
9446---> 5175--->exist
9447---> 5265--->exist
9448---> 5355--->exist
9449---> 5445--->exist
9450---> 9081--->exist
9451---> 8082--->exist
9452---> 7083--->exist
9453---> 6084--->exist
9454---> 5085--->exist
9455---> 4995--->exist
9456---> 5085--->exist
9457---> 5175--->exist
9458---> 5265--->exist
9459---> 5355--->exist
9460---> 9171--->exist
9461---> 8172--->exist
9462---> 7173--->exist
9463---> 61

## Магические числа, найденные для чисел разной разрядности (k), разных систем счислений (r): 
r=2 k=2 	MAGIC NUMBERS:  [0, 1]  
r=2 k=3 	MAGIC NUMBERS:  [0, 3]  
r=2 k=4 	MAGIC NUMBERS:  [0, 7, 9]  
r=2 k=5 	MAGIC NUMBERS:  [0, 15, 21]  
r=2 k=6 	MAGIC NUMBERS:  [0, 31, 45]  
r=2 k=7 	MAGIC NUMBERS:  [0, 63, 93]  
r=2 k=8 	MAGIC NUMBERS:  [0, 127, 189, 217]  
r=2 k=9 	MAGIC NUMBERS:  [0, 255, 381, 441]  
r=2 k=10 	MAGIC NUMBERS:  [0, 511, 765, 889]  
r=2 k=11 	MAGIC NUMBERS:  [0, 1023, 1533, 1785]  
r=2 k=12 	MAGIC NUMBERS:  [0, 2047, 3069, 3577]  
r=2 k=13 	MAGIC NUMBERS:  [0, 4095, 6141, 7161]  
r=2 k=14 	MAGIC NUMBERS:  [0, 8191, 12285, 14329]  
r=2 k=15 	MAGIC NUMBERS:  [0, 16383, 24573, 28665]  
r=2 k=16 	MAGIC NUMBERS:  [0, 32767, 49149, 57337, 61425]  
r=4 k=2 	MAGIC NUMBERS:  [0]  
r=4 k=3 	MAGIC NUMBERS:  [0, 30]  
r=4 k=4 	MAGIC NUMBERS:  [0, 201]  
r=4 k=5 	MAGIC NUMBERS:  [0]  
r=4 k=6 	MAGIC NUMBERS:  [0, 3369, 2550, 3873]  
r=4 k=7 	MAGIC NUMBERS:  [0, 14565]  
r=4 k=8 	MAGIC NUMBERS:  [0, 54441, 62625]  
r=4 k=9 	MAGIC NUMBERS:  [0, 234405, 171990, 254865]  
r=4 k=10 	MAGIC NUMBERS:  [0, 954261, 873129]  
r=4 k=11 	MAGIC NUMBERS:  [0, 3755685, 4083345]  
r=4 k=12 	MAGIC NUMBERS:  [0, 15285909, 13978281, 16075425]  
r=4 k=13 	MAGIC NUMBERS:  [0, 60111525, 61406805, 65354385, 66665025]  
r=4 k=14 	MAGIC NUMBERS:  [0, 223685289, 244644501, 262470225, 265628289]  
r=4 k=15 	MAGIC NUMBERS:  [0, 961866405, 982776405, 1050933585, 1066723905]  
r=4 k=16 	MAGIC NUMBERS:  [0, 3914590869, 3579095721, 3935304021, 4271106369]  
r=6 k=2 	MAGIC NUMBERS:  [0]  
r=6 k=3 	MAGIC NUMBERS:  [0, 105]  
r=6 k=4 	MAGIC NUMBERS:  [0]  
r=6 k=5 	MAGIC NUMBERS:  [0, 5600]  
r=6 k=6 	MAGIC NUMBERS:  [0, 33860, 42925, 27195]  
r=6 k=7 	MAGIC NUMBERS:  [0]  
r=6 k=8 	MAGIC NUMBERS:  [0, 1275170, 1657225]  
r=6 k=9 	MAGIC NUMBERS:  [0, 6018495]  
r=6 k=10 	MAGIC NUMBERS:  [0, 56319925, 45962330, 47681900]  
r=6 k=11 	MAGIC NUMBERS:  [0, 277695950, 348285175]  
r=6 k=12 	MAGIC NUMBERS:  [0, 2151904825, 1305060855]  
r=6 k=13 	MAGIC NUMBERS:  [0, 10363227560, 12973622725]  
r=6 k=14 	MAGIC NUMBERS:  [0, 73115587525, 60063064790, 59994427550, 78214899025]  
r=6 k=15 	MAGIC NUMBERS:  [0, 451814966575, 469662556825]  
r=6 k=16 	MAGIC NUMBERS:  [0, 2724011240875, 2789618448025]  
r=8 k=2 	MAGIC NUMBERS:  [0, 21]  
r=8 k=3 	MAGIC NUMBERS:  [0, 252]  
r=8 k=4 	MAGIC NUMBERS:  [0]  
r=8 k=5 	MAGIC NUMBERS:  [0]  
r=8 k=6 	MAGIC NUMBERS:  [0, 213402, 147420]  
r=8 k=7 	MAGIC NUMBERS:  [0, 1711962]  
r=8 k=8 	MAGIC NUMBERS:  [0, 16092433]  
r=8 k=9 	MAGIC NUMBERS:  [0, 76545756]  
r=8 k=10 	MAGIC NUMBERS:  [0, 1068263553]  
r=8 k=11 	MAGIC NUMBERS:  [0]  
r=8 k=12 	MAGIC NUMBERS:  [0, 57497839826, 58573445322, 39258683100]  
r=8 k=13 	MAGIC NUMBERS:  [0]  
r=8 k=14 	MAGIC NUMBERS:  [0, 4238480197777, 3749808363210, 3680015145170]  
r=8 k=15 	MAGIC NUMBERS:  [0, 34466323678857, 20104741893852]  
r=8 k=16 	MAGIC NUMBERS:  [0, 280198446204033]  
r=10 k=2 	MAGIC NUMBERS:  [0]  
r=10 k=3 	MAGIC NUMBERS:  [0, 495]  
r=10 k=4 	MAGIC NUMBERS:  [0, 6174]  
r=10 k=5 	MAGIC NUMBERS:  [0]  
r=10 k=6 	MAGIC NUMBERS:  [0, 631764, 549945]  
r=10 k=7 	MAGIC NUMBERS:  [0]  
r=10 k=8 	MAGIC NUMBERS:  [0, 97508421, 63317664]  
r=10 k=9 	MAGIC NUMBERS:  [0, 864197532, 554999445]  
r=10 k=10 	MAGIC NUMBERS:  [0, 9753086421, 9975084201, 6333176664]  
r=10 k=11 	MAGIC NUMBERS:  [0, 86431976532]  
r=10 k=12 	MAGIC NUMBERS:  [0, 975330866421, 555499994445, 997530864201, 999750842001, 633331766664]  
r=10 k=13 	MAGIC NUMBERS:  [0, 8643319766532]  
r=10 k=14 	MAGIC NUMBERS:  [0, 97533308666421, 97755108844221]  
r=10 k=15 	MAGIC NUMBERS:  [0, 864333197666532, 555549999944445]  
r=10 k=16 	MAGIC NUMBERS:  [0, 9775531088644221, 9977551088442201]  
r=12 k=2 	MAGIC NUMBERS:  [0]  
r=12 k=3 	MAGIC NUMBERS:  [0, 858]  
r=12 k=4 	MAGIC NUMBERS:  [0]  
r=12 k=5 	MAGIC NUMBERS:  [0, 172744]  
r=12 k=6 	MAGIC NUMBERS:  [0, 1617330]  
r=12 k=7 	MAGIC NUMBERS:  [0, 28428543]  
r=12 k=8 	MAGIC NUMBERS:  [0, 383468294, 308551012]  
r=12 k=9 	MAGIC NUMBERS:  [0, 2812795986]  
r=12 k=10 	MAGIC NUMBERS:  [0, 44873359960]  
r=12 k=11 	MAGIC NUMBERS:  [0, 663181025078, 534610567348]  
r=12 k=12 	MAGIC NUMBERS:  [0, 8768726887129, 4863092970450]  
r=12 k=13 	MAGIC NUMBERS:  [0, 77728651058344]  
r=12 k=14 	MAGIC NUMBERS:  [0, 1039052057510775, 1048030075321515, 924013465340020, 1282149981815329]  
r=12 k=15 	MAGIC NUMBERS:  [0, 8403796176530898, 11201862503984152]  
r=12 k=16 	MAGIC NUMBERS:  [0, 134342105155721320]  
r=14 k=2 	MAGIC NUMBERS:  [0, 65]  
r=14 k=3 	MAGIC NUMBERS:  [0, 1365]  
r=14 k=4 	MAGIC NUMBERS:  [0]  
r=14 k=5 	MAGIC NUMBERS:  [0]  
r=14 k=6 	MAGIC NUMBERS:  [0, 4033575]  
r=14 k=7 	MAGIC NUMBERS:  [0]  
r=14 k=8 	MAGIC NUMBERS:  [0, 1344056324]  
r=14 k=9 	MAGIC NUMBERS:  [0, 17116239465, 11121123195]  
r=14 k=10 	MAGIC NUMBERS:  [0]  
r=14 k=11 	MAGIC NUMBERS:  [0]  
r=14 k=12 	MAGIC NUMBERS:  [0, 30526696602675]  
r=14 k=13 	MAGIC NUMBERS:  [0, 723266716456440]  
r=14 k=14  

## Функция для вычисления максимальной длины цикла до нахождения магического числа или до зацикливания:

In [9]:
def max_cycle_length(k: int, r: int):
    count1 = 0 
    count2 = 0
    new = 0
    magic_nums = list()
    exist_nums = list()
    for i in range(10000):
        while True:
            new = kaprekar(i, k, r)
            if new == i:
                if new in magic_nums:
                    break
                magic_nums.append(new) 
            elif new in exist_nums:
                break
            i = new
            exist_nums.append(new)            
            count2+=1
        if count2>count1:
            count1 = count2
        count2 = 0
    return count1     

In [19]:
#Перевод из десятичной системы счисления в любую другую
def from_digital(number: int, k: int, r: int):
    list_of_digits = list()
    num = number
    for m in range(k):
        num, tmp = divmod(num, r)
        list_of_digits.append(tmp)
    list_of_digits.reverse()
    l = int("".join(map(str, list_of_digits)))  
    return l
    

In [10]:
for r in range (2, 16 + 1, 2):
    for k in range(2, 16 + 1):
        print("r = ", r, end='')
        print("k = ", k, end='')
        print("\t\t\tMAX_CYCLE_LENGTH: %d" % max_cycle_length(k,r))

r =  2k =  2			MAX_CYCLE_LENGTH: 1
r =  2k =  3			MAX_CYCLE_LENGTH: 2
r =  2k =  4			MAX_CYCLE_LENGTH: 2
r =  2k =  5			MAX_CYCLE_LENGTH: 2
r =  2k =  6			MAX_CYCLE_LENGTH: 2
r =  2k =  7			MAX_CYCLE_LENGTH: 2
r =  2k =  8			MAX_CYCLE_LENGTH: 2
r =  2k =  9			MAX_CYCLE_LENGTH: 2
r =  2k =  10			MAX_CYCLE_LENGTH: 2
r =  2k =  11			MAX_CYCLE_LENGTH: 2
r =  2k =  12			MAX_CYCLE_LENGTH: 2
r =  2k =  13			MAX_CYCLE_LENGTH: 2
r =  2k =  14			MAX_CYCLE_LENGTH: 2
r =  2k =  15			MAX_CYCLE_LENGTH: 2
r =  2k =  16			MAX_CYCLE_LENGTH: 2
r =  4k =  2			MAX_CYCLE_LENGTH: 2
r =  4k =  3			MAX_CYCLE_LENGTH: 4
r =  4k =  4			MAX_CYCLE_LENGTH: 4
r =  4k =  5			MAX_CYCLE_LENGTH: 3
r =  4k =  6			MAX_CYCLE_LENGTH: 7
r =  4k =  7			MAX_CYCLE_LENGTH: 6
r =  4k =  8			MAX_CYCLE_LENGTH: 8
r =  4k =  9			MAX_CYCLE_LENGTH: 6
r =  4k =  10			MAX_CYCLE_LENGTH: 6
r =  4k =  11			MAX_CYCLE_LENGTH: 6
r =  4k =  12			MAX_CYCLE_LENGTH: 7
r =  4k =  13			MAX_CYCLE_LENGTH: 7
r =  4k =  14			MAX_CYCLE_LENGTH: 7
r =  4k 

# Численный эксперемент:
## Задача: 
### Для чисел системы счисления 'r' (от 2 до 10), раздрядности 'k' (от 2 до 6) найти все магические числа и определить максимальную длину цикла

In [94]:
def numerical_experiment():
    statistics = dict()
    statistics = {(0, 0): 0}
    magic = tuple()
    nums = tuple()
    for r in range(2,10+1,1):
        for k in range(2,6+1,1):
            print("r=%d"%r,"k=%d"%k, end='') 
            magic = magic_numbers(k,r)
            print(" \tMAGIC NUMBERS: ", str(magic).ljust(25), end='')
            for i in range(len(magic)):
                magic[i]=from_digital(magic[i],k,r)
            print(' Original number system==({l})'.format(l=magic))
            magic.append(str(max_cycle_length(k,r)))
            statistics[(r, k)]= magic            
    return statistics  

        

In [95]:
numerical_experiment()


r=2 k=2 	MAGIC NUMBERS:  [0, 1]                    Original number system==([0, 1])
r=2 k=3 	MAGIC NUMBERS:  [0, 3]                    Original number system==([0, 11])
r=2 k=4 	MAGIC NUMBERS:  [0, 7, 9]                 Original number system==([0, 111, 1001])
r=2 k=5 	MAGIC NUMBERS:  [0, 15, 21]               Original number system==([0, 1111, 10101])
r=2 k=6 	MAGIC NUMBERS:  [0, 31, 45, 49]           Original number system==([0, 11111, 101101, 110001])
r=3 k=2 	MAGIC NUMBERS:  [0]                       Original number system==([0])
r=3 k=3 	MAGIC NUMBERS:  [0]                       Original number system==([0])
r=3 k=4 	MAGIC NUMBERS:  [0]                       Original number system==([0])
r=3 k=5 	MAGIC NUMBERS:  [0, 184]                  Original number system==([0, 20211])
r=3 k=6 	MAGIC NUMBERS:  [0]                       Original number system==([0])
r=4 k=2 	MAGIC NUMBERS:  [0]                       Original number system==([0])
r=4 k=3 	MAGIC NUMBERS:  [0, 30]                

{(0, 0): 0,
 (2, 2): [0, 1, '1'],
 (2, 3): [0, 11, '2'],
 (2, 4): [0, 111, 1001, '2'],
 (2, 5): [0, 1111, 10101, '2'],
 (2, 6): [0, 11111, 101101, 110001, '2'],
 (3, 2): [0, '2'],
 (3, 3): [0, '2'],
 (3, 4): [0, '3'],
 (3, 5): [0, 20211, '5'],
 (3, 6): [0, '4'],
 (4, 2): [0, '2'],
 (4, 3): [0, 132, '4'],
 (4, 4): [0, 3021, '4'],
 (4, 5): [0, '3'],
 (4, 6): [0, 310221, 213312, 330201, '7'],
 (5, 2): [0, 13, '4'],
 (5, 3): [0, '4'],
 (5, 4): [0, 3032, '4'],
 (5, 5): [0, '6'],
 (5, 6): [0, '7'],
 (6, 2): [0, '3'],
 (6, 3): [0, 253, '5'],
 (6, 4): [0, '8'],
 (6, 5): [0, 41532, '7'],
 (6, 6): [0, 420432, 530421, 325523, '6'],
 (7, 2): [0, '3'],
 (7, 3): [0, '5'],
 (7, 4): [0, '5'],
 (7, 5): [0, '10'],
 (7, 6): [0, '11'],
 (8, 2): [0, 25, '3'],
 (8, 3): [0, 374, '6'],
 (8, 4): [0, '7'],
 (8, 5): [0, '8'],
 (8, 6): [0, 640632, 437734, '10'],
 (9, 2): [0, '4'],
 (9, 3): [0, '6'],
 (9, 4): [0, '5'],
 (9, 5): [0, 62853, '9'],
 (9, 6): [0, '16'],
 (10, 2): [0, '5'],
 (10, 3): [0, 495, '7'],
 (10,

## Значения statistics:  
### формат - (k,r) : [*nums, 'max_length']  
#### k-разрядность  
#### r-система счисления  
#### *nums - список магических точек  
#### max_length - максимальная длина цикла  
{(0, 0): 0,  
 (2, 2): [0, 1, '1'],  
 (2, 3): [0, 11, '2'],  
 (2, 4): [0, 111, 1001, '2'],  
 (2, 5): [0, 1111, 10101, '2'],  
 (2, 6): [0, 11111, 101101, '2'],  
 (3, 2): [0, '2'],  
 (3, 3): [0, '2'],  
 (3, 4): [0, '3'],  
 (3, 5): [0, 20211, '5'],  
 (3, 6): [0, '4'],  
 (4, 2): [0, '2'],  
 (4, 3): [0, 132, '4'],  
 (4, 4): [0, 3021, '4'],  
 (4, 5): [0, '3'],  
 (4, 6): [0, 310221, 213312, 330201, '7'],  
 (5, 2): [0, 13, '4'],  
 (5, 3): [0, '4'],  
 (5, 4): [0, 3032, '4'],  
 (5, 5): [0, '6'],  
 (5, 6): [0, '7'],  
 (6, 2): [0, '3'],  
 (6, 3): [0, 253, '5'],  
 (6, 4): [0, '8'],  
 (6, 5): [0, 41532, '7'],  
 (6, 6): [0, 420432, 530421, 325523, '6'],  
 (7, 2): [0, '3'],  
 (7, 3): [0, '5'],  
 (7, 4): [0, '5'],  
 (7, 5): [0, '10'],  
 (7, 6): [0, '11'],  
 (8, 2): [0, 25, '3'],  
 (8, 3): [0, 374, '6'],  
 (8, 4): [0, '7'],  
 (8, 5): [0, '8'],  
 (8, 6): [0, 640632, 437734, '10'],  
 (9, 2): [0, '4'],  
 (9, 3): [0, '6'],  
 (9, 4): [0, '5'],  
 (9, 5): [0, 62853, '9'],  
 (9, 6): [0, '16'],  
 (10, 2): [0, '5'],  
 (10, 3): [0, 495, '7'],  
 (10, 4): [0, 6174, '6'],  
 (10, 5): [0, '9'],  
 (10, 6): [0, 631764, 549945, '17']}  

# Результаты численного эксперемента
### Полученные данные представлены в таблице
строки таблицы - разрядность числа, колонки - система счисления  
результаты вычислений отображаются в виде [*magic_numbers_list, max_cycle_length]

In [85]:
import pandas as pd
import numpy as np
a = np.zeros((9,5), dtype = list)
s = numerical_experiment()
for key in s:
    a[key[0]-2][key[1]-2]=s[key]
    if key[0] is not 0:
        tmp=a[key[0]-2][key[1]-2][-1]
        del a[key[0]-2][key[1]-2][-1]
        a[key[0]-2][key[1]-2].append("max_cycle = %s"%tmp)      
        
table = pd.DataFrame(a,columns = [str(r) + " digit" for r in range(2, 7)], index = [str(k) + " radix" for k in range(2, 11)])

table

2 digit                  3 digit  \
2 radix    [0, 1, max_cycle = 1]   [0, 11, max_cycle = 2]   
3 radix       [0, max_cycle = 2]       [0, max_cycle = 2]   
4 radix       [0, max_cycle = 2]  [0, 132, max_cycle = 4]   
5 radix   [0, 13, max_cycle = 4]       [0, max_cycle = 4]   
6 radix       [0, max_cycle = 3]  [0, 253, max_cycle = 5]   
7 radix       [0, max_cycle = 3]       [0, max_cycle = 5]   
8 radix   [0, 25, max_cycle = 3]  [0, 374, max_cycle = 6]   
9 radix       [0, max_cycle = 4]       [0, max_cycle = 6]   
10 radix      [0, max_cycle = 5]  [0, 495, max_cycle = 7]   

                                4 digit                          5 digit  \
2 radix   [0, 111, 1001, max_cycle = 2]  [0, 1111, 10101, max_cycle = 2]   
3 radix              [0, max_cycle = 3]        [0, 20211, max_cycle = 5]   
4 radix        [0, 3021, max_cycle = 4]               [0, max_cycle = 3]   
5 radix        [0, 3032, max_cycle = 4]               [0, max_cycle = 6]   
6 radix              [0, max_cycle = 8]        [0, 41532, max_cycle = 7]   
7 radix              [0, max_cycle = 5]              [0, max_cycle = 10]   
8 radix              [0, max_cycle = 7]               [0, max_cycle = 8]   
9 radix              [0, max_cycle = 5]        [0, 62853, max_cycle = 9]   
10 radix       [0, 6174, max_cycle = 6]               [0, max_cycle = 9]   

                                             6 digit  
2 radix            [0, 11111, 101101, max_cycle = 2]  
3 radix                           [0, max_cycle = 4]  
4 radix   [0, 310221, 213312, 330201, max_cycle = 7]  
5 radix                           [0, max_cycle = 7]  
6 radix   [0, 420432, 530421, 325523, max_cycle = 6]  
7 radix                          [0, max_cycle = 11]  
8 radix          [0, 640632, 437734, max_cycle = 10]  
9 radix                          [0, max_cycle = 16]  
10 radix         [0, 631764, 549945, max_cycle = 17]